crates/uv-resolver/src/candidate_selector.rs

```diff
@@ -324,7 +324,31 @@ impl CandidateSelector {
             version_maps.iter().map(VersionMap::len).sum::<usize>(),
         );
         let highest = self.use_highest_version(package_name);
-        let allow_prerelease = self.prerelease_strategy.allows(package_name, markers);
+
+        let allow_prerelease = match self.prerelease_strategy.allows(package_name, markers) {
+            AllowPrerelease::Yes => AllowPrerelease::Yes,
+            AllowPrerelease::No => AllowPrerelease::No,
+            AllowPrerelease::IfNecessary => {
+                // If we're looking at a pre-release singleton, pre-check whether _all_ versions are
+                // pre-release.
+                if range
+                    .as_singleton()
+                    .is_some_and(|version| version.any_prerelease())
+                {
+                    if version_maps.iter().any(|version_map| {
+                        version_map
+                            .versions()
+                            .all(|version| version.any_prerelease())
+                    }) {
+                        AllowPrerelease::Yes
+                    } else {
+                        AllowPrerelease::No
+                    }
```

Comment:
A little bit of a penalty here but should be somewhat rare.

Reply:
We only need to do this slow path when we see a pre-release singleton; and the check will short-circuit as soon as we see any non-pre-release.

Reply:
It'd be nice to avoid this though.

Reply:
You could store whether all versions are prereleases in the version map, probably helpful for debugging anyway

Reply:
Good idea.